src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Hierarchy.cs

```diff
@@ -184,405 +192,206 @@ public static partial class SymbolFinder
             if (solution.GetOriginatingProjectId(type) == null)
                 throw new ArgumentException(WorkspacesResources.Symbols_project_could_not_be_found_in_the_provided_solution, nameof(type));
 
-            return await DependentTypeFinder.FindTransitivelyDerivedClassesAsync(
-                type, solution, projects, cancellationToken).ConfigureAwait(false);
+            return await DependentTypeFinder.FindAndCacheDerivedClassesAsync(
+                type, solution, projects, transitive: true, cancellationToken).ConfigureAwait(false);
         }
 
         /// <summary>
-        /// Finds the symbols that implement an interface or interface member.
+        /// Finds the immediate derived classes of the given type. Implementations of an interface are not considered
+        /// "derived", but can be found with <see cref="FindImplementationsAsync(ISymbol, Solution,
+        /// IImmutableSet{Project}, CancellationToken)"/>.
         /// </summary>
-        public static async Task<IEnumerable<ISymbol>> FindImplementationsAsync(
-            ISymbol symbol, Solution solution, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default)
-        {
-            if (solution.GetOriginatingProjectId(symbol) == null)
-                throw new ArgumentException(WorkspacesResources.Symbols_project_could_not_be_found_in_the_provided_solution, nameof(symbol));
-
-            return await FindImplementationsArrayAsync(symbol, solution, projects, cancellationToken).ConfigureAwait(false);
-        }
-
-        internal static async Task<ImmutableArray<ISymbol>> FindImplementationsArrayAsync(
-            ISymbol symbol, Solution solution, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default)
+        /// <param name="type">The symbol to find derived types of.</param>
+        /// <param name="solution">The solution to search in.</param>
+        /// <param name="projects">The projects to search. Can be null to search the entire solution.</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>The derived types of the symbol. The symbol passed in is not included in this list.</returns>
+        public static async Task<IEnumerable<INamedTypeSymbol>> FindImmediateDerivedClassesAsync(
+            INamedTypeSymbol type, Solution solution, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default)
```

Comment:
for all the existing transitive helpers, we also now expose the immediate form.  This way tools can be build that can either show full hierarchies, or can expand one level on demand.    The nameing is:

`FindWhateverAsync` for the transntive form, and `FindImmediateWhateverAsync` for the one-level form.  The reason i don't put `Transitive' in the name of the former is that we've already shipped several, so we can't change the naming here.

Another option for this public api would be to have the following:

```c#
 public static async Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
            INamedTypeSymbol type, Solution solution, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default)
    => FindDerivedClassesAsync(type, solution, transitive: true, projects, cancellationToken);

public static async Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
            INamedTypeSymbol type, Solution solution, bool transitive, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default) => ...
```

I'm open to the latter if people prefer it.